### ISSUE TITLE
Invalid frame length when used with Kodi (aka: XBMC)

### ISSUE THREAD

**Issue by technicalflaw** [tidwall/SwiftWebSocket#7]
I'm noticing my JSON-RPC messages aren't getting processed when I use SwiftWebSocket to send remote control messages to Kodi. The error log for Kodi states the following:

```
15:49:34 T:4582928384    INFO: WebSocket: Frame with invalid length received
15:49:34 T:4582928384    INFO: WebSocket: Invalid frame received
15:49:34 T:4582928384    INFO: WebSocket: Frame with invalid length received
15:49:34 T:4582928384    INFO: WebSocket: Invalid frame received
15:49:34 T:4582928384    INFO: WebSocket: Fragmented control frame (opcode  9) received
15:49:34 T:4582928384    INFO: WebSocket: Invalid frame received
15:49:34 T:4582928384    INFO: WebSocket: Frame with invalid length received
15:49:34 T:4582928384    INFO: WebSocket: Invalid frame received
```

It seems to work once occasionally, though. I'm assuming it's some sort of formatting/checksum error on one end of the connection. No unusual config for SwiftWebSocket. BinaryType is .UInt8Array. Any ideas?

**Comment by tidwall** [comment]
Sorry to hear that you're having issues with the library.

This is the first I've heard of this type of issue. I've run SwiftWebSocket through the Autobahn test suite which performs hundreds of random tests relating framing and fragmentation, so I'm certainly curious if you found a bug.

Are you running both endpoints on the same local network? Is there a proxy in the middle? 

What I would do to determine if SwiftWebSocket is the cause is to send the same messages to an echo endpoint (such as wss://echo.websocket.org) and see if they come back intact. There's an example in the README. Running over WSS/HTTPS will ensure that no proxies will get in the way.

Feel free to email me over some/many failing message payloads and I can do a deep dive.

**Comment by technicalflaw** [comment]
OK, I've done a bit more investigation. Both endpoints are on the same network (localhost), with no proxy. I'm using Xcode 7b5 with iOS 9 (simulator), and Kodi 15.0. Here's the ViewController of a Single Page app to test it (everything else is as per the template. A button is the source of the action):

```
import UIKit
import SwiftWebSocket

class ViewController: UIViewController {
    let ws = WebSocket("ws://localhost:9090/jsonrpc")

    override func viewDidLoad() {
        super.viewDidLoad()

        self.ws.event.open = { print("didOpen") }
        self.ws.event.close = { _ in print("didClose") }
        self.ws.event.message = { print("message: \($0)") }
    }

    @IBAction func didTouchUpInside(sender: AnyObject) {
        // basic jsonrpc command (doesn't seem to matter in this case)
        self.ws.send("{ \"jsonrpc\": \"2.0\", \"method\": \"JSONRPC.Version\", \"id\": 1 }")
    }
}
```

Using `ws://echo.websocket.org` it works as expected, but with Kodi a handful of responses will arrive, until the aforementioned frame errors show in the Kodi's console log, which it doesn't recover from, and eventually causes the connection to be forcibly closed, though I don't know by which end of the connection. Interestingly I haven't noticed the same issue with Starscream.

If you'd like a pcap file or something like that, get in touch

**Comment by tidwall** [comment]
This is good info. I'll mirror your environment locally to reproduce the issue and let you know.

**Comment by tidwall** [comment]
I was able to reproduce the issue. Though I can't exactly pin down the culprit. SwiftWebSocket and Kodi both seem to be sending and receiving the exact same messages successfully when I run them though a small proxy that I wrote in Go. 

There's a SwiftWebSocket 'dev' branch which contains a massive performance boost and a smaller memory footprint. It also so happens to fix this issue. 

If you would have time to test the 'dev' branch with your app, I would very much appreciate it. If you do see success I'll wrap it up and merge it asap.

Thanks a ton for your help.
